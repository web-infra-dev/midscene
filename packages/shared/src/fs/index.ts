import assert from 'node:assert';
import { existsSync, readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';

interface PkgInfo {
  name: string;
  version: string;
  dir: string;
}

let pkg: PkgInfo | undefined;
export function getMidscenePkgInfo(dir: string): PkgInfo {
  if (pkg) {
    return pkg;
  }

  const pkgDir = findNearestPackageJson(dir || process.cwd());
  assert(pkgDir, 'package.json not found');
  const pkgJsonFile = join(pkgDir, 'package.json');

  if (pkgJsonFile) {
    const { name, version } = JSON.parse(readFileSync(pkgJsonFile, 'utf-8'));
    pkg = { name, version, dir: pkgDir };
    return pkg;
  }
  return {
    name: 'midscene-unknown-page-name',
    version: '0.0.0',
    dir: pkgDir,
  };
}

/**
 * Find the nearest package.json file recursively
 * @param {string} dir - Home directory
 * @returns {string|null} - The most recent package.json file path or null
 */
export function findNearestPackageJson(dir: string): string | null {
  const packageJsonPath = join(dir, 'package.json');
  if (existsSync(packageJsonPath)) {
    return dir;
  }

  const parentDir = dirname(dir);

  // Return null if the root directory has been reached
  if (parentDir === dir) {
    return null;
  }

  return findNearestPackageJson(parentDir);
}
