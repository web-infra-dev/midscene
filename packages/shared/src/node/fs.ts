import { existsSync, readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { ifInBrowser, ifInWorker } from '../utils';

declare const __HTML_ELEMENT_SCRIPT__: string;

interface PkgInfo {
  name: string;
  version: string;
  dir: string;
}

const pkgCacheMap: Record<string, PkgInfo> = {};

export function getRunningPkgInfo(dir?: string): PkgInfo | null {
  if (ifInBrowser || ifInWorker) {
    return null;
  }
  const dirToCheck = dir || process.cwd();
  if (pkgCacheMap[dirToCheck]) {
    return pkgCacheMap[dirToCheck];
  }

  const pkgDir = findNearestPackageJson(dirToCheck);
  const pkgJsonFile = pkgDir ? join(pkgDir, 'package.json') : null;

  if (pkgDir && pkgJsonFile) {
    const { name, version } = JSON.parse(readFileSync(pkgJsonFile, 'utf-8'));
    pkgCacheMap[dirToCheck] = {
      name: name || 'midscene-unknown-package-name',
      version: version || '0.0.0',
      dir: pkgDir,
    };
    return pkgCacheMap[dirToCheck];
  }
  return {
    name: 'midscene-unknown-package-name',
    version: '0.0.0',
    dir: dirToCheck,
  };
}

/**
 * Find the nearest package.json file recursively
 * @param {string} dir - Home directory
 * @returns {string|null} - The most recent package.json file path or null
 */
export function findNearestPackageJson(dir: string): string | null {
  const packageJsonPath = join(dir, 'package.json');
  if (existsSync(packageJsonPath)) {
    return dir;
  }

  const parentDir = dirname(dir);

  // Return null if the root directory has been reached
  if (parentDir === dir) {
    return null;
  }

  return findNearestPackageJson(parentDir);
}

export function getElementInfosScriptContent() {
  const htmlElementScript = __HTML_ELEMENT_SCRIPT__;

  if (!htmlElementScript) {
    throw new Error('HTML_ELEMENT_SCRIPT inject failed.');
  }
  return htmlElementScript;
}

export async function getExtraReturnLogic(tree = false) {
  if (ifInBrowser || ifInWorker) {
    return null;
  }

  const elementInfosScriptContent = `${getElementInfosScriptContent()}midscene_element_inspector.setNodeHashCacheListOnWindow();`;

  if (tree) {
    return `${elementInfosScriptContent}midscene_element_inspector.webExtractNodeTree()`;
  }
  return `${elementInfosScriptContent}midscene_element_inspector.webExtractTextWithPosition()`;
}
