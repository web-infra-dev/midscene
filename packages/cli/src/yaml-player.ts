import { createServer } from 'http-server';
import yaml from 'js-yaml';
import puppeteer from 'puppeteer';
import type { Browser, Page } from 'puppeteer';

import assert from 'node:assert';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { basename, dirname, extname, join } from 'node:path';
import { PuppeteerAgent } from '@midscene/web/puppeteer';
import { paramStr, typeStr } from '@midscene/web/ui-utils';

import {
  contextInfo,
  contextTaskListSummary,
  isTTY,
  singleTaskInfo,
  spinnerInterval,
} from './printer';
import { TTYWindowRenderer } from './tty-renderer';
import type {
  MidsceneYamlFileContext,
  MidsceneYamlFlowItemAIAction,
  MidsceneYamlFlowItemAIAssert,
  MidsceneYamlFlowItemAIQuery,
  MidsceneYamlFlowItemAIWaitFor,
  MidsceneYamlFlowItemSleep,
  MidsceneYamlScript,
  ScriptPlayerOptions,
  ScriptPlayerStatusValue,
  ScriptPlayerTaskStatus,
} from './types';

export const defaultUA =
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36';
export const defaultViewportWidth = 1280;
export const defaultViewportHeight = 960;
export const defaultViewportScale = process.platform === 'darwin' ? 2 : 1;
export const defaultWaitForNetworkIdleTimeout = 10 * 1000;

export function loadYamlScript(
  content: string,
  filePath?: string,
): MidsceneYamlScript {
  const obj = yaml.load(content) as MidsceneYamlScript;
  const pathTip = filePath ? `, failed to load ${filePath}` : '';
  assert(obj.target, `property "target" is required in yaml script${pathTip}`);
  assert(
    typeof obj.target === 'object',
    `property "target" must be an object${pathTip}`,
  );
  assert(
    typeof obj.target.url === 'string',
    `property "target.url" must be provided in yaml script: ${pathTip}`,
  );
  assert(obj.tasks, `property "tasks" is required in yaml script${pathTip}`);
  assert(
    Array.isArray(obj.tasks),
    `property "tasks" must be an array${pathTip}`,
  );
  return obj;
}

export const launchServer = async (
  dir: string,
): Promise<ReturnType<typeof createServer>> => {
  // https://github.com/http-party/http-server/blob/master/bin/http-server
  return new Promise((resolve, reject) => {
    const server = createServer({
      root: dir,
    });
    server.listen(0, '127.0.0.1', () => {
      resolve(server);
    });
  });
};

let ttyRenderer: TTYWindowRenderer | undefined;
export async function playYamlFiles(
  files: string[],
  options?: ScriptPlayerOptions,
): Promise<boolean> {
  // prepare
  const fileContextList: MidsceneYamlFileContext[] = [];
  for (const file of files) {
    const script = loadYamlScript(readFileSync(file, 'utf-8'), file);
    const fileName = basename(file, extname(file));
    const player = new ScriptPlayer(script, {
      ...options,
      testId: fileName,
      onTaskStatusChange: (taskStatus) => {
        if (!isTTY) {
          const { nameText } = singleTaskInfo(taskStatus);
          // console.log(`${taskStatus.status} - ${nameText}`);
        }
      },
    });
    fileContextList.push({ file, player });
  }

  // play
  if (isTTY) {
    const summaryContents = () => {
      const summary: string[] = [''];
      for (const context of fileContextList) {
        summary.push(
          contextTaskListSummary(context.player.taskStatus, context),
        );
      }
      summary.push('');
      return summary;
    };
    ttyRenderer = new TTYWindowRenderer({
      outputStream: process.stdout,
      errorStream: process.stderr,
      getWindow: summaryContents,
      interval: spinnerInterval,
    });

    ttyRenderer.start();
    for (const context of fileContextList) {
      await context.player.run();
    }
    ttyRenderer.stop();
  } else {
    for (const context of fileContextList) {
      const { mergedText } = contextInfo(context);
      console.log(mergedText);
      await context.player.run();
      console.log(contextTaskListSummary(context.player.taskStatus, context));
    }
  }

  const ifFail = fileContextList.some((task) => task.player.status === 'error');
  return !ifFail;
}

export class ScriptPlayer {
  public currentTaskIndex?: number;
  public taskStatus: ScriptPlayerTaskStatus[] = [];
  public status: ScriptPlayerStatusValue = 'init';
  public reportFile?: string | null;
  public result: Record<string, any>;
  private unnamedResultIndex = 0;
  public output?: string | null;
  public errorInSetup?: Error;
  constructor(
    private script: MidsceneYamlScript,
    private options?: ScriptPlayerOptions,
  ) {
    this.result = {};
    this.output = script.target.output;
    this.taskStatus = (script.tasks || []).map((task, taskIndex) => ({
      ...task,
      index: taskIndex,
      status: 'init',
      totalSteps: task.flow?.length || 0,
    }));
  }

  private setPlayerStatus(status: ScriptPlayerStatusValue, error?: Error) {
    this.status = status;
    this.errorInSetup = error;
  }

  private setTaskStatus(
    index: number,
    statusValue: ScriptPlayerStatusValue,
    error?: Error,
  ) {
    this.taskStatus[index].status = statusValue;
    if (error) {
      this.taskStatus[index].error = error;
    }

    if (this.options?.onTaskStatusChange) {
      this.options.onTaskStatusChange(this.taskStatus[index]);
    }
  }

  private setTaskIndex(taskIndex: number) {
    this.currentTaskIndex = taskIndex;
  }

  private flushResult() {
    if (Object.keys(this.result).length && this.output) {
      const output = join(process.cwd(), this.output);
      // ensure the output directory exists
      const outputDir = dirname(output);
      if (!existsSync(outputDir)) {
        mkdirSync(outputDir, { recursive: true });
      }
      writeFileSync(output, JSON.stringify(this.result, undefined, 2));
    }
  }

  async playTask(
    taskStatus: ScriptPlayerTaskStatus,
    {
      agent,
      browser,
      page,
    }: {
      agent: PuppeteerAgent;
      browser: Browser;
      page: Page;
    },
  ) {
    const { flow } = taskStatus;
    assert(flow, 'missing flow in task');

    for (const flowItemIndex in flow) {
      taskStatus.currentStep = Number.parseInt(flowItemIndex, 10);
      const flowItem = flow[flowItemIndex];
      if (
        (flowItem as MidsceneYamlFlowItemAIAction).aiAction ||
        (flowItem as MidsceneYamlFlowItemAIAction).ai
      ) {
        const actionTask = flowItem as MidsceneYamlFlowItemAIAction;
        const prompt = actionTask.aiAction || actionTask.ai;
        assert(prompt, 'missing prompt for ai (aiAction)');
        assert(
          typeof prompt === 'string',
          'prompt for aiAction must be a string',
        );
        await agent.aiAction(prompt, {
          onTaskStart(task) {
            const tip = `${typeStr(task)} - ${paramStr(task)}`;
            (flowItem as MidsceneYamlFlowItemAIAction).aiActionProgressTip =
              tip;
          },
        });
      } else if ((flowItem as MidsceneYamlFlowItemAIAssert).aiAssert) {
        const assertTask = flowItem as MidsceneYamlFlowItemAIAssert;
        const prompt = assertTask.aiAssert;
        assert(prompt, 'missing prompt for aiAssert');
        assert(
          typeof prompt === 'string',
          'prompt for aiAssert must be a string',
        );
        await agent.aiAssert(prompt);
      } else if ((flowItem as MidsceneYamlFlowItemAIQuery).aiQuery) {
        const queryTask = flowItem as MidsceneYamlFlowItemAIQuery;
        const prompt = queryTask.aiQuery;
        assert(prompt, 'missing prompt for aiQuery');
        assert(
          typeof prompt === 'string',
          'prompt for aiQuery must be a string',
        );
        const queryResult = await agent.aiQuery(prompt);
        const resultKey = queryTask.name || this.unnamedResultIndex++;
        if (this.result[resultKey]) {
          console.warn(
            `result key ${resultKey} already exists, will overwrite`,
          );
        }

        this.result[resultKey] = queryResult;
        this.flushResult();
      } else if ((flowItem as MidsceneYamlFlowItemAIWaitFor).aiWaitFor) {
        const waitForTask = flowItem as MidsceneYamlFlowItemAIWaitFor;
        const prompt = waitForTask.aiWaitFor;
        assert(prompt, 'missing prompt for aiWaitFor');
        assert(
          typeof prompt === 'string',
          'prompt for aiWaitFor must be a string',
        );
        const timeout = waitForTask.timeout;
        await agent.aiWaitFor(prompt, { timeoutMs: timeout });
      } else if ((flowItem as MidsceneYamlFlowItemSleep).sleep) {
        const sleepTask = flowItem as MidsceneYamlFlowItemSleep;
        const ms = sleepTask.sleep;
        assert(
          ms && ms > 0,
          `ms for sleep must be greater than 0, but got ${ms}`,
        );
        await new Promise((resolve) => setTimeout(resolve, ms));
      } else {
        throw new Error(`unknown flowItem: ${JSON.stringify(flowItem)}`);
      }
    }
    this.reportFile = agent.reportFile;
  }

  async run() {
    const { target, tasks } = this.script;
    this.setPlayerStatus('running');

    // prepare the environment
    const ua = target.userAgent || defaultUA;
    let width = defaultViewportWidth;
    if (target.viewportWidth) {
      assert(
        typeof target.viewportWidth === 'number',
        'viewportWidth must be a number',
      );
      width = Number.parseInt(target.viewportWidth as unknown as string, 10);
      assert(
        width > 0,
        `viewportWidth must be greater than 0, but got ${width}`,
      );
    }
    let height = defaultViewportHeight;
    if (target.viewportHeight) {
      assert(
        typeof target.viewportHeight === 'number',
        'viewportHeight must be a number',
      );
      height = Number.parseInt(target.viewportHeight as unknown as string, 10);
      assert(
        height > 0,
        `viewportHeight must be greater than 0, but got ${height}`,
      );
    }
    let dpr = defaultViewportScale;
    if (target.viewportScale) {
      assert(
        typeof target.viewportScale === 'number',
        'viewportScale must be a number',
      );
      dpr = Number.parseInt(target.viewportScale as unknown as string, 10);
      assert(dpr > 0, `viewportScale must be greater than 0, but got ${dpr}`);
    }
    const viewportConfig = {
      width,
      height,
      deviceScaleFactor: dpr,
    };

    const freeFn: {
      name: string;
      fn: () => void;
    }[] = [];

    let localServer: Awaited<ReturnType<typeof launchServer>> | undefined;
    let urlToVisit: string | undefined;
    assert(typeof target.url === 'string', 'url is required');
    if (target.serve) {
      localServer = await launchServer(target.serve);
      const serverAddress = localServer.server.address();
      freeFn.push({
        name: 'local_server',
        fn: () => localServer?.server.close(),
      });
      if (target.url.startsWith('/')) {
        urlToVisit = `http://${serverAddress?.address}:${serverAddress?.port}${target.url}`;
      } else {
        urlToVisit = `http://${serverAddress?.address}:${serverAddress?.port}/${target.url}`;
      }
    } else {
      urlToVisit = target.url;
    }

    const headed = this.options?.headed || this.options?.keepWindow;
    // launch the browser
    if (headed && process.env.CI === '1') {
      console.warn(
        'you are probably running headed mode in CI, this will usually fail.',
      );
    }
    const browser = await puppeteer.launch({
      headless: !headed,
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-features=PasswordLeakDetection',
        '--disable-save-password-bubble',
      ],
    });
    freeFn.push({
      name: 'puppeteer_browser',
      fn: () => {
        if (!this.options?.keepWindow) {
          browser.close();
        }
      },
    });

    const pages = await browser.pages();
    const page = pages[0];
    await page.setUserAgent(ua);
    await page.setViewport(viewportConfig);

    if (target.cookie) {
      const cookieFileContent = readFileSync(target.cookie, 'utf-8');
      await page.setCookie(...JSON.parse(cookieFileContent));
    }

    await page.goto(urlToVisit);
    const waitForNetworkIdleTimeout =
      typeof target.waitForNetworkIdle?.timeout === 'number'
        ? target.waitForNetworkIdle.timeout
        : defaultWaitForNetworkIdleTimeout;
    try {
      if (waitForNetworkIdleTimeout > 0) {
        await page.waitForNetworkIdle({
          timeout: waitForNetworkIdleTimeout,
        });
      }
    } catch (e) {
      if (
        typeof target.waitForNetworkIdle?.continueOnNetworkIdleError ===
          'boolean' &&
        !target.waitForNetworkIdle?.continueOnNetworkIdleError
      ) {
        const newError = new Error(`failed to wait for network idle: ${e}`, {
          cause: e,
        });
        this.setPlayerStatus('error', newError);
        return;
      }
      const newMessage = `failed to wait for network idle after ${waitForNetworkIdleTimeout}ms, but the script will continue.`;
      console.warn(newMessage);
    }

    // prepare Midscene agent
    const agent = new PuppeteerAgent(page, {
      autoPrintReportMsg: false,
      testId: this.options?.testId,
    });

    freeFn.push({
      name: 'midscene_puppeteer_agent',
      fn: () => agent.destroy(),
    });

    let taskIndex = 0;
    this.setPlayerStatus('running');
    let errorFlag = false;
    while (taskIndex < tasks.length) {
      const task = tasks[taskIndex];
      this.setTaskStatus(taskIndex, 'running' as any);
      // const taskStatus = this.taskStatus[taskIndex];

      try {
        this.setTaskIndex(taskIndex);
        await this.playTask(this.taskStatus[taskIndex], {
          agent,
          browser,
          page,
        });
      } catch (e) {
        this.setTaskStatus(taskIndex, 'error' as any, e as Error);
        this.setPlayerStatus('error');
        errorFlag = true;

        this.reportFile = agent.reportFile;
        taskIndex++;
        continue;
      }
      this.reportFile = agent.reportFile;
      this.setTaskStatus(taskIndex, 'done' as any);
      taskIndex++;
    }

    if (!errorFlag) {
      this.setPlayerStatus('done');
    }

    // free the resources
    freeFn.forEach((fn) => {
      try {
        fn.fn();
      } catch (e) {}
    });
  }
}
