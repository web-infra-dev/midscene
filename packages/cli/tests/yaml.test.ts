import { randomUUID } from 'node:crypto';
import { existsSync } from 'node:fs';
import { join } from 'node:path';
import { ScriptPlayer, launchServer, loadYamlScript } from '@/yaml-player';
import { assert, describe, expect, test, vi } from 'vitest';

const runYaml = async (yamlString: string) => {
  const script = loadYamlScript(yamlString);
  const player = new ScriptPlayer(script);
  await player.run();
  assert(
    player.status === 'done',
    player.errorInSetup?.message || 'unknown error',
  );
};

const shouldRunAITest =
  process.platform !== 'linux' || process.env.AITEST === 'true';

const serverRoot = join(__dirname, 'server_root');
describe('yaml', () => {
  test('basic load', () => {
    const script = loadYamlScript(`
      target:
        url: https://www.baidu.com
        waitForNetworkIdle:
          timeout: 1000
          continueOnNetworkIdleError: true
      tasks:
        - name: search
          flow:
            - action: type 'hello' in search box, hit enter
    `);

    expect(script).toMatchSnapshot();
  });

  test('load error with filePath', () => {
    expect(() => {
      loadYamlScript(
        `
      target:
        a: 1
      `,
        'some_error_path',
      );
    }).toThrow(/some_error_path/);
  });

  test('launch server', async () => {
    const serverResult = await launchServer(serverRoot);
    expect(serverResult).toBeDefined();

    const serverAddress = serverResult.server.address();
    const staticServerUrl = `http://${serverAddress?.address}:${serverAddress?.port}`;

    const contents = await fetch(`${staticServerUrl}/index.html`);
    expect(contents.status).toBe(200);

    await serverResult.server.close();
  });

  test('player - bad params', async () => {
    expect(async () => {
      await runYaml(`
          target:
            serve: ${serverRoot}
        `);
    }).rejects.toThrow();

    expect(async () => {
      await runYaml(`
          target:
            serve: ${serverRoot}
            viewportWidth: 0
        `);
    }).rejects.toThrow();
  });
});

describe.skipIf(!shouldRunAITest)(
  'player - e2e',
  () => {
    test('local server', async () => {
      const yamlString = `
      target:
        serve: ${serverRoot}
        url: index.html
        viewportWidth: 300
        viewportHeight: 500
      tasks:
        - name: local page
          flow:
            - aiAssert: the content title is "My App"
    `;

      await runYaml(yamlString);
    });

    test('local server - flush output even if assertion failed', async () => {
      const outputPath = `./midscene_run/output/${randomUUID()}.json`;
      const yamlString = `
      target:
        serve: ${serverRoot}
        url: index.html
        output: ${outputPath}
      tasks:
        - name: local page
          flow:
            - aiQuery: >
                the background color of the page, { color: 'white' | 'black' | 'red' | 'green' | 'blue' | 'yellow' | 'purple' | 'orange' | 'pink' | 'brown' | 'gray' | 'black' 
        - name: check content
          flow:
            - aiAssert: this is a search result page
      `;
      await expect(async () => {
        await runYaml(yamlString);
      }).rejects.toThrow();

      expect(existsSync(outputPath)).toBe(true);
    });

    test('local server - assertion failed', async () => {
      const yamlString = `
      target:
        serve: ${serverRoot}
        url: index.html
        viewportWidth: 300
        viewportHeight: 500
      tasks:
        - name: check content
          flow:
            - aiAssert: it shows the width is 888
      `;

      expect(async () => {
        await runYaml(yamlString);
      }).rejects.toThrow();
    });

    test('cookie', async () => {
      const yamlString = `
      target:
        url: http://httpbin.dev/cookies
        cookie: ./tests/fixture/httpbin.dev_cookies.json
      tasks:
        - name: check cookie
          flow:
            - aiAssert: the value of midscene_foo is "bar"
    `;
      await runYaml(yamlString);
    });

    test('online server - lazy response', async () => {
      const yamlString = `
      target:
        url: https://httpbin.org/delay/60000
        waitForNetworkIdle:
          timeout: 10
          continueOnNetworkIdleError: false
      tasks:
        - name: check content
          flow:
            - aiAssert: the response is "Hello, world!"
    `;

      expect(async () => {
        await runYaml(yamlString);
      }).rejects.toThrow(/TimeoutError/i);
    });
  },
  60 * 1000,
);
