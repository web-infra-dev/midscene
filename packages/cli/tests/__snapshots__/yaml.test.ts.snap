// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`yaml > basic load 1`] = `
{
  "target": {
    "url": "https://www.baidu.com",
    "waitForNetworkIdle": {
      "continueOnNetworkIdleError": true,
      "timeout": 1000,
    },
  },
  "tasks": [
    {
      "flow": [
        {
          "action": "type 'hello' in search box, hit enter",
        },
      ],
      "name": "search",
    },
  ],
}
`;
